### ISSUE TITLE
Providers loaded by FileSystemProvider not handled correctly in native image

### ISSUE THREAD

**Issue by OlgaMaciaszek** [oracle/graal#5134]
Providers loaded by `java.nio.file.spi.FileSystemProvider` (such as `BouncyCastleProvider` or `SftpFileSystemProvider`) cause native image compilation to fail with `No instances of xxx are allowed in the image heap as this class should be initialized at image runtime.`. However, adding `--initialize-at-run-time=java.nio.file.spi.FileSystemProvider` to setup results in `Error: The class java.nio.file.spi.FileSystemProvider has already been initialized (from command line with 'java.nio.file.spi.FileSystemProvider'); it is too late to register java.nio.file.spi.FileSystemProvider for build-time initialization.`.

You can use [this demo branch](https://github.com/OlgaMaciaszek/demo-config-server-native/tree/simple-setup) as a sample. 
Run with `./mvnw clean package -Pnative -DskipTests`.

**Comment by vjovanov** [comment]
This requires moving FileSystemProviders to runtime. I tried making this change and many things failed because of the initialization strategy we use. I would wait for the new initialization strategy before addressing this issue.

**Comment by OlgaMaciaszek** [comment]
Thanks, @vjovanov. What would the timeline be for this?

**Comment by mikehearn** [comment]
I'm hitting this too. It's annoying because the FS provider is brought in via transitive dependencies.

**Comment by bkesmarki** [comment]
I'd also be interested in any news about the issue. 
We are using a custom EnvironmentRepository, but from the result/symptoms perspective, we are blocked by the same error with graalvm, native image execution.

**Comment by gopalharis** [comment]
We are in need of the fix for this issue, this is the only blocker we have for graalvm native build

**Comment by sschuberth** [comment]
> such as `BouncyCastleProvider` or `SftpFileSystemProvider`

FYI, in case of `SftpFileSystemProvider` from Apache MINA, coming in as a transitive dependency of JGit, I was able to work around the issue by simply excluding the dependency, because SFTP is usually not needed as a Git transport:

    implementation("org.eclipse.jgit:org.eclipse.jgit.ssh.apache.agent:6.7.0.202309050840-r") {
        exclude(group = "org.apache.sshd", module = "sshd-sftp")
            .because("it is not required for cloning via SSH and causes issues with GraalVM native images")
    }

**Comment by avollmaier** [comment]
I also need a fix for this bug - blocker

**Comment by dudiao** [comment]
When will it be repaired?

**Comment by vjovanov** [comment]
Would setting up the `-H:-AddAllFileSystemProviders` at image build help as a temporary workaround.

**Comment by OlgaMaciaszek** [comment]
@vjovanov - have tried that, but got a different issue then:

**Comment by vjovanov** [comment]
This looks like a message from older versions of GraalVM. Have you tried with the latest dev build, or with the GraalVM for JDK 21 with a flag `--strict-image-heap`?

**Comment by OlgaMaciaszek** [comment]
That was from `17.0.9`. Now from `21.0.1`, getting `Error: Classes that should be initialized at run time got initialized during image building` followed by a lot of classes, big part of them BounceCastle-related. Adding `--initialize-at-build-time` options has fixed it. However, there's now another issue: during initialization of BouncyCastle classes, we get issues related to `SP800SecureRandom` being initialized at build time, but then  if we set `--initialize-at-run-time=org.bouncycastle.jcajce.provider.drbg.DRBG` (that being the class the instantiation can be traced to), then we get:
```
 The class org.bouncycastle.jcajce.provider.drbg.DRBG has already been initialized (from '[property source]' with 'org.bouncycastle.jcajce.provider.drbg.DRBG'); it is too late to register org.bouncycastle.jcajce.provider.drbg.DRBG for build-time initialization.
```

**Comment by vjovanov** [comment]
I think we need to add the option `--strict-image-heap` to make it work.

**Comment by OlgaMaciaszek** [comment]
That would be those snapshots; but I'll update the sample build config with snapshot repos today and let you know - sorry for missing that. Also, I have managed to make a native SC Config server work with [this workaround](https://github.com/oracle/graal/issues/5134#issuecomment-1784572396) as a native image for some basic scenarios. Testing some more scenarios.

**Comment by OlgaMaciaszek** [comment]
So, actually, after switching to [this setup](https://github.com/spring-cloud/spring-cloud-config/blob/native-image-for-server/spring-cloud-config-server/src/main/resources/META-INF/native-image/org.springframework.cloud/spring-cloud-config-server/native-image.properties) and fixing `java.security` providers config (I think there's an issue in the doc, have created a PR with fix suggestion, waiting for approval to accept CLI: https://github.com/oracle/graal/pull/7855), I was able to run it with `-H:-AddAllFileSystemProviders`, without the exclusion. I guess a reproducer is not exactly needed at this point, but I'm adding updated info for documentation purposes: SC Config  [ branch](https://github.com/spring-cloud/spring-cloud-config/tree/native-image-for-server) of SC Config (`./mvnw clean install -Pspring` to build), then use it to create and run a native image on [this demo branch](https://github.com/OlgaMaciaszek/demo-config-server-native/tree/security-ssh) (`./mvnw native:compile -Pnative` and `./target/demo-config-server-ref`) . Thanks for your help @vjovanov .

**Comment by vjovanov** [comment]
Hope this can be a workaround for everyone until this issue is fixed as a part of https://github.com/oracle/graal/issues/7488.

**Comment by OlgaMaciaszek** [comment]
It does work, however, since it requires[ initialising `logback`-related classes at build time](https://github.com/spring-cloud/spring-cloud-config/blob/native-image-for-server/spring-cloud-config-server/src/main/resources/META-INF/native-image/org.springframework.cloud/spring-cloud-config-server/native-image.properties) to work, and after chatting with the team, the decision is we'll not be able to support it, so still a blocker. I will take another look at the exclusion workaround and post here.

**Comment by vjovanov** [comment]
Have you tried to reduce that list of initialized classes when the `--strict-image-heap` is used?

**Comment by OlgaMaciaszek** [comment]
Thanks for the suggestion @vjovanov. With `--strict-image-heap` and `-H:-AddAllFileSystemProviders` (without the exclude), I was able to make a native image Config server with git backend over ssh work with only `EdDSASecurityProvider` and the org.bouncycastle` components initialised at build time (setup [here](https://github.com/spring-cloud/spring-cloud-config/blob/native-image-for-server/spring-cloud-config-server/src/main/resources/META-INF/native-image/org.springframework.cloud/spring-cloud-config-server/native-image.properties))  - I think there's no  way around it since any security providers need to be initialised then anyhow? Will be testing some other scenarios now.

**Comment by vjovanov** [comment]
We will try to move security classes to run-time init in the future, but it is uncertain whether it will be possible. Until then this looks like a good solution.

**Comment by KafkaProServerless** [comment]
Impacted by this

**Comment by andreiyusupau** [comment]
@OlgaMaciaszek maybe a stupid question, but where exactly args like `-H:-AddAllFileSystemProvider` should be added? I use gradle with following config, but it doesn't really seems to help:
``` groovy
graalvmNative {
    binaries {
        main {
            buildArgs.add('-H:-AddAllFileSystemProviders')
            buildArgs.add('--strict-image-heap')
            }
            }
            }
            ```

**Comment by OlgaMaciaszek** [comment]
@andreiyusupau  It does work correctly in our sample setup: https://github.com/spring-projects/spring-aot-smoke-tests/blob/main/cloud/cloud-config-server/build.gradle.  You need to make sure you have the right version of Graal (21 or higher).

**Comment by zapster** [comment]
We have now support for initializing `FileSystemProvider`s at image run time (#11205). Since this is a major behavioral change, it needs to be unlocked with `--future-defaults=...` (see #11205 for more details). I encourage everyone facing a related problem to try it out and report issues if there are any. The more bugs we find now, the more fixes can make it into the upcoming GraalVM for JDK 25 LTS release. :)

**Comment by yu7863956** [comment]
> We have now support for initializing `FileSystemProvider`s at image run time ([#11205](https://github.com/oracle/graal/pull/11205)). Since this is a major behavioral change, it needs to be unlocked with `--future-defaults=...` (see [#11205](https://github.com/oracle/graal/pull/11205) for more details). I encourage everyone facing a related problem to try it out and report issues if there are any. The more bugs we find now, the more fixes can make it into the upcoming GraalVM for JDK 25 LTS release. :)

Could you please tell me if this repair is effective for this version of GraalVM JDK 17.0.16 (Linux x64)?
When I use it this way, the prompt shows "Error: Unrecognized option: --future-defaults=run-time-initialized-jdk"
native-image
-cp "$CLASSPATH"
--no-fallback
--initialize-at-build-time=kotlin.jvm.internal
-H:Name="$OUTPUT"
-H:+ReportExceptionStackTraces
-H:-AddAllFileSystemProviders
--future-defaults=run-time-initialized-jdk
--initialize-at-run-time=org.apache.sshd
--trace-class-initialization=org.apache.sshd.sftp.client.fs.SftpFileSystemProvider
-H:+AllowIncompleteClasspath
--report-unsupported-elements-at-runtime
--delay-class-initialization-to-runtime=org.apache.sshd.sftp.client.fs.SftpFileSystemProvider
-H:ReflectionConfigurationFiles=/home/lumi/native/reflect-config.json
com.lumi.framework.KotlinSpringbootApplicationKt

**Comment by wirthi** [comment]
@yu7863956 this fix will only land in the (not yet releases) JDK 25 version of GraalVM (and later).

@zapster can you give some guidance what can be done instead on GraalVM JDK 17.0.16?

**Comment by zapster** [comment]
> [@zapster](https://github.com/zapster) can you give some guidance what can be done instead on GraalVM JDK 17.0.16?

Nothing that I know of. The code base of 17 is very different and has other usability problems as well.
@yu7863956 any reason why you need to stick with JDK 17?

**Comment by yu7863956** [comment]
> > [@zapster](https://github.com/zapster) can you give some guidance what can be done instead on GraalVM JDK 17.0.16?
> 
> Nothing that I know of. The code base of 17 is very different and has other usability problems as well. [@yu7863956](https://github.com/yu7863956) any reason why you need to stick with JDK 17?

We are upgrading to JDK 21 and have added these two configurations: -H: -AddAllFileSystemProviders --strict-image-heap. This issue has been resolved, but we have encountered some other problems. I would like to confirm if this modification only takes effect in JDK 25? Thanks your answer.

**Comment by zapster** [comment]
> I would like to confirm if this modification only takes effect in JDK 25?

yes, no plans to backport it to any other supported version.
